### PR TITLE
feat(forge): replace `TypedTransaction` with `TransactionWithMetadata` on `ScriptSequence`

### PR DESCRIPTION
**Pull request by joshieDo**
## Motivation

Needed a better way to map what happens on the script to the broadcast log.

## Solution
When running the onchain simulation we collect the available metadata and store it alongside the transaction in a new type `TransactionWithMetadata`. It actually ends up simplifying certain things (eg. create2 handling) in other parts of the code. Also moved `ScriptSequence` into its own module.

Looking for feedback on what to include and/or remove.

### Example

<details>
  <summary>broadcast.log</summary>

```json
{
    "transactions": [
        {
            "hash": "0xa49dbb4843ab3ae83b72178b8dc3feaba9d949f48006e3c0db3d93b7dac07703",
            "type": "CREATE",
            "contractName": "Contract",
            "contractAddress": "0x0xc86cf989b79e97bfe039e28c0b12efdaefc6222f",
            "arguments": [],
            "tx": {
                "type": "0x02",
                "from": "0x31a5aa527cd8879d01728b320751fced3776c703",
                "gas": "0x2cdba",
                "value": "0x0",
                "data": "0x608060405234801561001057600080fd5b50610198806100206000396000f3fe608060405234801561001057600080fd5b50600436106100365760003560e01c806306b0a5bc1461003b5780630d10a4eb1461004f575b600080fd5b600060405190815260200160405180910390f35b61006261005d36600461007a565b505050565b005b634e487b7160e01b600052604160045260246000fd5b60008060006060848603121561008f57600080fd5b833592506020808501356001600160a01b03811681146100ae57600080fd5b9250604085013567ffffffffffffffff808211156100cb57600080fd5b818701915087601f8301126100df57600080fd5b8135818111156100f1576100f1610064565b8060051b604051601f19603f8301168101818110858211171561011657610116610064565b60405291825284820192508381018501918a83111561013457600080fd5b938501935b8285101561015257843584529385019392850192610139565b809650505050505050925092509256fea2646970667358221220d38e358fd29f0dc5a81b2fd401a0c41049fe495f8d9e996a59f11f7eb69b5be164736f6c634300080d0033",
                "nonce": "0xc17",
                "accessList": []
            }
        },
        {
            "hash": "0x3dead4ab28110b00bbabb775353728e09a7c5e73e597e19d4988ff7972756c73",
            "type": "CREATE2",
            "contractName": "Contract",
            "contractAddress": "0x0x5d3f0ff987c47064cc92b7b35047cc58606c6c86",
            "arguments": [],
            "tx": {
                "type": "0x02",
                "from": "0x31a5aa527cd8879d01728b320751fced3776c703",
                "to": "0x4e59b44847b379578588920ca78fbf26c0b4956c",
                "gas": "0x2cfe2",
                "value": "0x0",
                "data": "0x0000000000000000000000000000000000000000000000000000000062a230a2608060405234801561001057600080fd5b50610198806100206000396000f3fe608060405234801561001057600080fd5b50600436106100365760003560e01c806306b0a5bc1461003b5780630d10a4eb1461004f575b600080fd5b600060405190815260200160405180910390f35b61006261005d36600461007a565b505050565b005b634e487b7160e01b600052604160045260246000fd5b60008060006060848603121561008f57600080fd5b833592506020808501356001600160a01b03811681146100ae57600080fd5b9250604085013567ffffffffffffffff808211156100cb57600080fd5b818701915087601f8301126100df57600080fd5b8135818111156100f1576100f1610064565b8060051b604051601f19603f8301168101818110858211171561011657610116610064565b60405291825284820192508381018501918a83111561013457600080fd5b938501935b8285101561015257843584529385019392850192610139565b809650505050505050925092509256fea2646970667358221220d38e358fd29f0dc5a81b2fd401a0c41049fe495f8d9e996a59f11f7eb69b5be164736f6c634300080d0033",
                "nonce": "0xc18",
                "accessList": []
            }
        },
        {
            "hash": "0x626bff91b8c430ef2a2b1d5fbf36fd62ccaaba2ae992a33d579cd1e1ea7ff39a",
            "type": "CALL",
            "contractName": "Test",
            "contractAddress": "0x0x43b4d30bb88fb260e2d142c2c350305c2d84df25",
            "function": "inc():(uint256)",
            "arguments": [],
            "tx": {
                "type": "0x02",
                "from": "0x31a5aa527cd8879d01728b320751fced3776c703",
                "to": "0x43b4d30bb88fb260e2d142c2c350305c2d84df25",
                "gas": "0x85f8",
                "value": "0x0",
                "data": "0x371303c0",
                "nonce": "0xc19",
                "accessList": []
            }
        },
        {
            "hash": "0x84ba0a5deeac500013f6b5b61e41395b290568f04b0111f7ba971164b032a2e8",
            "type": "CALL",
            "contractName": "Test",
            "contractAddress": "0x0x43b4d30bb88fb260e2d142c2c350305c2d84df25",
            "function": "t(uint256):(uint256)",
            "arguments": [
                "2"
            ],
            "tx": {
                "type": "0x02",
                "from": "0x31a5aa527cd8879d01728b320751fced3776c703",
                "to": "0x43b4d30bb88fb260e2d142c2c350305c2d84df25",
                "gas": "0x9b2a",
                "value": "0x0",
                "data": "0xafe29f710000000000000000000000000000000000000000000000000000000000000002",
                "nonce": "0xc1a",
                "accessList": []
            }
        },
        {
            "hash": "0x57fcefb29af3b685e649ee5225bb8af754cec6683b719804e79dd8006bfb0680",
            "type": "CALL",
            "contractName": "Contract",
            "contractAddress": "0x0xc86cf989b79e97bfe039e28c0b12efdaefc6222f",
            "function": "hello(uint256,address,uint256[])",
            "arguments": [
                "1",
                "0000000000000000000000000000000000001337",
                "[3,4]"
            ],
            "tx": {
                "type": "0x02",
                "from": "0x31a5aa527cd8879d01728b320751fced3776c703",
                "to": "0xc86cf989b79e97bfe039e28c0b12efdaefc6222f",
                "gas": "0x733b",
                "value": "0x0",
                "data": "0x0d10a4eb000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000013370000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000030000000000000000000000000000000000000000000000000000000000000004",
                "nonce": "0xc1b",
                "accessList": []
            }
        },
        {
            "hash": "0x384d5dcb817900d9d0a531af06571df052ee0bf62fb22a5694bb1447875dd296",
            "type": "CALL",
            "contractName": "Contract",
            "contractAddress": "0x0x5d3f0ff987c47064cc92b7b35047cc58606c6c86",
            "function": "hello(uint256,address,uint256[])",
            "arguments": [
                "1",
                "0000000000000000000000000000000000001337",
                "[3,4]"
            ],
            "tx": {
                "type": "0x02",
                "from": "0x31a5aa527cd8879d01728b320751fced3776c703",
                "to": "0x5d3f0ff987c47064cc92b7b35047cc58606c6c86",
                "gas": "0x733b",
                "value": "0x0",
                "data": "0x0d10a4eb000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000013370000000000000000000000000000000000000000000000000000000000000060000000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000030000000000000000000000000000000000000000000000000000000000000004",
                "nonce": "0xc1c",
                "accessList": []
            }
        },
        {
            "hash": "0xcb5c2472165c2a6a1c03b6f71a045ae829c998c10b2b420930115a1af5de89da",
            "type": "CALL",
            "arguments": [],
            "tx": {
                "type": "0x02",
                "from": "0x31a5aa527cd8879d01728b320751fced3776c703",
                "to": "0x0000000000000000000000000000000000001337",
                "gas": "0x6aa4",
                "value": "0x5af3107a4000",
                "data": "0x",
                "nonce": "0xc1d",
                "accessList": []
            }
        },
        {
            "hash": "0x3fe5a87d04589e4bd6801e67fe9a9caa7e854c1f215edacec341c29747bfc99b",
            "type": "CALL",
            "contractName": "Contract",
            "contractAddress": "0x0xc86cf989b79e97bfe039e28c0b12efdaefc6222f",
            "function": "no_args():(uint256)",
            "arguments": [],
            "tx": {
                "type": "0x02",
                "from": "0x31a5aa527cd8879d01728b320751fced3776c703",
                "to": "0xc86cf989b79e97bfe039e28c0b12efdaefc6222f",
                "gas": "0x6bb5",
                "value": "0x0",
                "data": "0x06b0a5bc",
                "nonce": "0xc1e",
                "accessList": []
            }
        }
    ],
    "receipts": [
        {
            "transactionHash": "0xa49dbb4843ab3ae83b72178b8dc3feaba9d949f48006e3c0db3d93b7dac07703",
            "transactionIndex": "0x0",
            "blockHash": "0x164d3037870be56f0b1c98fe6997b6179b5244110e94a3f44659a09809d29bc0",
            "blockNumber": "0xa13b8d",
            "from": "0x31a5aa527cd8879d01728b320751fced3776c703",
            "to": null,
            "cumulativeGasUsed": "0x22819",
            "gasUsed": "0x22819",
            "contractAddress": "0xc86cf989b79e97bfe039e28c0b12efdaefc6222f",
            "logs": [],
            "status": "0x1",
            "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
            "type": "0x2",
            "effectiveGasPrice": "0x684ee1800"
        },
        {
            "transactionHash": "0x3dead4ab28110b00bbabb775353728e09a7c5e73e597e19d4988ff7972756c73",
            "transactionIndex": "0x1",
            "blockHash": "0x164d3037870be56f0b1c98fe6997b6179b5244110e94a3f44659a09809d29bc0",
            "blockNumber": "0xa13b8d",
            "from": "0x31a5aa527cd8879d01728b320751fced3776c703",
            "to": "0x4e59b44847b379578588920ca78fbf26c0b4956c",
            "cumulativeGasUsed": "0x451db",
            "gasUsed": "0x229c2",
            "contractAddress": null,
            "logs": [],
            "status": "0x1",
            "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
            "type": "0x2",
            "effectiveGasPrice": "0x684ee1800"
        },
        {
            "transactionHash": "0x626bff91b8c430ef2a2b1d5fbf36fd62ccaaba2ae992a33d579cd1e1ea7ff39a",
            "transactionIndex": "0x1",
            "blockHash": "0x9c10966b4b3fd0443a2850417ffe85b67bd9372b485ce12567149c16fe6f6d42",
            "blockNumber": "0xa13b8e",
            "from": "0x31a5aa527cd8879d01728b320751fced3776c703",
            "to": "0x43b4d30bb88fb260e2d142c2c350305c2d84df25",
            "cumulativeGasUsed": "0x3d050",
            "gasUsed": "0x670e",
            "contractAddress": null,
            "logs": [],
            "status": "0x1",
            "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
            "type": "0x2",
            "effectiveGasPrice": "0x684ee1800"
        },
        {
            "transactionHash": "0x84ba0a5deeac500013f6b5b61e41395b290568f04b0111f7ba971164b032a2e8",
            "transactionIndex": "0x0",
            "blockHash": "0x29125cc95feb01f4e14d8bd2900a19b64b3a5d3ca6f169c0cd4a6863bd1fd4c5",
            "blockNumber": "0xa13b90",
            "from": "0x31a5aa527cd8879d01728b320751fced3776c703",
            "to": "0x43b4d30bb88fb260e2d142c2c350305c2d84df25",
            "cumulativeGasUsed": "0x775c",
            "gasUsed": "0x775c",
            "contractAddress": null,
            "logs": [
                {
                    "address": "0x43b4d30bb88fb260e2d142c2c350305c2d84df25",
                    "topics": [
                        "0x0b2e13ff20ac7b474198655583edf70dedd2c1dc980e329c4fbb2fc0748b796b"
                    ],
                    "data": "0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000046865726500000000000000000000000000000000000000000000000000000000",
                    "blockHash": "0x29125cc95feb01f4e14d8bd2900a19b64b3a5d3ca6f169c0cd4a6863bd1fd4c5",
                    "blockNumber": "0xa13b90",
                    "transactionHash": "0x84ba0a5deeac500013f6b5b61e41395b290568f04b0111f7ba971164b032a2e8",
                    "transactionIndex": "0x0",
                    "logIndex": "0x0",
                    "removed": false
                }
            ],
            "status": "0x1",
            "logsBloom": "0x00000000000000000000000000000010000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000900",
            "type": "0x2",
            "effectiveGasPrice": "0x684ee1800"
        },
        {
            "transactionHash": "0x57fcefb29af3b685e649ee5225bb8af754cec6683b719804e79dd8006bfb0680",
            "transactionIndex": "0x1",
            "blockHash": "0x29125cc95feb01f4e14d8bd2900a19b64b3a5d3ca6f169c0cd4a6863bd1fd4c5",
            "blockNumber": "0xa13b90",
            "from": "0x31a5aa527cd8879d01728b320751fced3776c703",
            "to": "0xc86cf989b79e97bfe039e28c0b12efdaefc6222f",
            "cumulativeGasUsed": "0xd000",
            "gasUsed": "0x58a4",
            "contractAddress": null,
            "logs": [],
            "status": "0x1",
            "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
            "type": "0x2",
            "effectiveGasPrice": "0x684ee1800"
        },
        {
            "transactionHash": "0x384d5dcb817900d9d0a531af06571df052ee0bf62fb22a5694bb1447875dd296",
            "transactionIndex": "0x0",
            "blockHash": "0x8aeb1366d36d596e2f71111e9bfdefb8726a5162e73d2ed712731f16da640fc1",
            "blockNumber": "0xa13b91",
            "from": "0x31a5aa527cd8879d01728b320751fced3776c703",
            "to": "0x5d3f0ff987c47064cc92b7b35047cc58606c6c86",
            "cumulativeGasUsed": "0x58a4",
            "gasUsed": "0x58a4",
            "contractAddress": null,
            "logs": [],
            "status": "0x1",
            "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
            "type": "0x2",
            "effectiveGasPrice": "0x684ee1800"
        },
        {
            "transactionHash": "0xcb5c2472165c2a6a1c03b6f71a045ae829c998c10b2b420930115a1af5de89da",
            "transactionIndex": "0x0",
            "blockHash": "0xebe8e3e3a31e88a52e09a503b107adb36d432263f1ecfc13be879b540467063f",
            "blockNumber": "0xa13b92",
            "from": "0x31a5aa527cd8879d01728b320751fced3776c703",
            "to": "0x0000000000000000000000000000000000001337",
            "cumulativeGasUsed": "0x5208",
            "gasUsed": "0x5208",
            "contractAddress": null,
            "logs": [],
            "status": "0x1",
            "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
            "type": "0x2",
            "effectiveGasPrice": "0x684ee1800"
        },
        {
            "transactionHash": "0x3fe5a87d04589e4bd6801e67fe9a9caa7e854c1f215edacec341c29747bfc99b",
            "transactionIndex": "0x0",
            "blockHash": "0x41766d26c79c56c4927dd200aad28e583de7c37783b9262cb897280d80488b30",
            "blockNumber": "0xa13b94",
            "from": "0x31a5aa527cd8879d01728b320751fced3776c703",
            "to": "0xc86cf989b79e97bfe039e28c0b12efdaefc6222f",
            "cumulativeGasUsed": "0x52da",
            "gasUsed": "0x52da",
            "contractAddress": null,
            "logs": [],
            "status": "0x1",
            "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
            "type": "0x2",
            "effectiveGasPrice": "0x684ee1800"
        }
    ],
    "libraries": [],
    "pending": [],
    "path": "broadcast/Contract.s.sol/43113/run-latest.json",
    "timestamp": 1654796512
}
  ```
</details>

### CODE REVIEW

**Review by gakonst — COMMENTED** [review]
Love it - we should add examples on how to consume these eg with ethers rs/js

**Comment by gakonst** [comment]
@mds1 wdyt of this output format?

**Comment by tynes** [comment]
> @mds1 wdyt of this output format (see the broadcast.log above)

`hardhat-deploy` creates deployment files with this schema: https://github.com/wighawag/hardhat-deploy/blob/819df0fad56d75a5de5218c3307bec2093f8794c/types.ts#L358

We have tooling that can then return a connected `ethers.Contract` given a network from one of these deployment artifacts.

I think going from an artifact to an initialized object that can talk to the contract in any language is the ideal outcome.

**Comment by joshieDo** [comment]
Opening for review. Still going to run a few more times to test.

File: cli/src/cmd/forge/script/broadcast.rs [hunk 1]
```diff
@@ -255,35 +253,33 @@ impl ScriptArgs {
...
+        txes: VecDeque<TransactionWithMetadata>,
         provider: Arc<Provider<RetryClient<Http>>>,
         chain: u64,
-    ) -> eyre::Result<VecDeque<TypedTransaction>> {
+    ) -> eyre::Result<VecDeque<TransactionWithMetadata>> {
         let is_legacy =
             self.legacy || Chain::try_from(chain).map(|x| Chain::is_legacy(&x)).unwrap_or_default();
 
-        let mut new_txes: VecDeque<TypedTransaction> = VecDeque::new();
+        let mut new_txes: VecDeque<TransactionWithMetadata> = VecDeque::new();
```
Comment (mattsse) [inline cli/src/cmd/forge/script/broadcast.rs:263]: do we need the type hint here?

File: cli/src/cmd/forge/script/sequence.rs [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+            target,
+            chain_id,
+        )?)?;
+        serde_json::from_str(&file).map_err(|e| e.into())
+    }
+
+    pub fn save(&mut self) -> eyre::Result<()> {
+        if !self.transactions.is_empty() {
+            self.timestamp =
+                SystemTime::now().duration_since(UNIX_EPOCH).expect("Wrong system time.").as_secs();
```
Comment (mattsse) [inline cli/src/cmd/forge/script/sequence.rs:80]: since we return a Result, return an err here instead

File: cli/src/cmd/forge/script/sequence.rs [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+
+            let path = self.path.to_str().expect("Invalid path.");
```
Comment (mattsse) [inline cli/src/cmd/forge/script/sequence.rs:82]: either lossy or return an error

File: cli/src/cmd/forge/script/sequence.rs [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+            serde_json::to_writer(
+                BufWriter::new(std::fs::File::create(
+                    &path.replace("latest.json", &format!("{}.json", self.timestamp)),
+                )?),
+                &self,
+            )?;
+
+            println!(
+                "\nTransactions saved to: {}\n",
+                self.path.to_str().expect(
```
Comment (mattsse) [inline cli/src/cmd/forge/script/sequence.rs:96]: same

File: cli/src/cmd/forge/script/sequence.rs [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+    /// Saves to ./broadcast/contract_filename/sig[-timestamp].json
+    pub fn get_path(
+        out: &Path,
+        sig: &str,
+        target: &ArtifactId,
+        chain_id: u64,
+    ) -> eyre::Result<PathBuf> {
+        let mut out = out.to_path_buf();
+
+        let target_fname = target.source.file_name().expect("No file name");
```
Comment (mattsse) [inline cli/src/cmd/forge/script/sequence.rs:153]: same

File: cli/src/cmd/forge/script/sequence.rs [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+        out.push(target_fname);
+        out.push(format!("{chain_id}"));
+
+        std::fs::create_dir_all(out.clone())?;
+
+        let filename = sig.split_once('(').expect("Sig is invalid").0.to_owned();
```
Comment (mattsse) [inline cli/src/cmd/forge/script/sequence.rs:159]: if the function returns an error return an error instead

File: cli/src/cmd/forge/script/sequence.rs [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+                a.transaction_index.cmp(&b.transaction_index)
+            } else {
+                ablock.cmp(&bblock)
+            }
+        });
+    }
+
+    pub fn add_pending(&mut self, index: usize, tx_hash: TxHash) {
+        if !self.pending.contains(&tx_hash) {
+            self.transactions.get_mut(index).expect("").hash = Some(tx_hash);
```
Comment (mattsse) [inline cli/src/cmd/forge/script/sequence.rs:123]: if this is save you can use `[index]` instead

File: cli/src/cmd/forge/script/sequence.rs [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+    }
+
+    pub fn add_libraries(&mut self, libraries: Libraries) {
+        self.libraries = {
+            let mut str_libs = vec![];
+            for (file, libs) in libraries.libs {
+                for (name, address) in libs {
+                    str_libs.push(format!("{}:{}:{}", file.to_string_lossy(), name, address));
+                }
+            }
```
Comment (mattsse) [inline cli/src/cmd/forge/script/sequence.rs:139]: you could try to replace this with a one liner via `flat_map` perhaps

File: cli/src/cmd/forge/script/sequence.rs [hunk 9]
```diff
@@ -0,0 +1,384 @@
...
+        target: &ArtifactId,
+        chain_id: u64,
+    ) -> eyre::Result<PathBuf> {
+        let mut out = out.to_path_buf();
+
+        let target_fname = target.source.file_name().expect("No file name");
+        out.push(target_fname);
+        out.push(format!("{chain_id}"));
+
+        std::fs::create_dir_all(out.clone())?;
```
Comment (mattsse) [inline cli/src/cmd/forge/script/sequence.rs:157]:
```suggestion
        std::fs::create_dir_all(&out))?;
```

**Comment by gakonst** [comment]
I dont think we should conform to any project's output format for this feature Fwiw

File: cli/src/cmd/forge/script/broadcast.rs [hunk 10]
```diff
@@ -255,35 +253,33 @@ impl ScriptArgs {
...
+
+            let typed_tx = tx.typed_tx_mut();
 
             if has_different_gas_calc(chain) {
-                tx.set_gas(provider.estimate_gas(&tx).await?);
+                typed_tx.set_gas(provider.estimate_gas(typed_tx).await?);
             }
 
-            total_gas += *tx.gas().expect("");
+            total_gas += *typed_tx.gas().expect("");
```
Comment (mattsse) [inline cli/src/cmd/forge/script/broadcast.rs:274]:
```suggestion
            total_gas += *typed_tx.gas().expect("gas is set");
```

File: cli/src/cmd/forge/script/sequence.rs [hunk 11]
```diff
@@ -0,0 +1,379 @@
...
+    pub fn add_libraries(&mut self, libraries: Libraries) {
+        self.libraries = libraries
+            .libs
+            .iter()
+            .flat_map(|(file, libs)| {
+                libs.iter().map(|(name, address)| {
+                    format!("{}:{}:{}", file.to_string_lossy(), name, address)
+                })
+            })
+            .collect();
```
Comment (mattsse) [inline cli/src/cmd/forge/script/sequence.rs:136]: nice

**Review by mattsse — CHANGES_REQUESTED** [review]
nice, only a couple of pedantic nits 🙃

**Comment by ShravanSunder** [comment]
@joshieDo it would be good to have a specification of the output file so we can use it in front end libraries that consume contract definitions. It’s similar to what we do with hardhat_contracts.json

File: cli/src/cmd/forge/script/executor.rs [hunk 12]
```diff
@@ -73,69 +71,72 @@ impl ScriptArgs {
...
+            .iter()
+            .filter_map(|(addr, contract_name)| {
+                if let Some((_, (abi, _))) =
+                    contracts.iter().find(|(artifact, _)| artifact.name == *contract_name)
+                {
+                    return Some((*addr, (contract_name.clone(), abi)))
+                }
+                None
+            })
+            .collect();
```
Comment (gakonst) [inline cli/src/cmd/forge/script/executor.rs:96]: Nit: we should make this a helper on the decoder

File: cli/src/cmd/forge/script/mod.rs [hunk 13]
```diff
@@ -426,3 +431,42 @@ pub struct ScriptConfig {
...
+}
+
+impl VerifyBundle {
+    pub fn new(
+        project: &Project,
+        config: &Config,
+        known_contracts: BTreeMap<ArtifactId, (Abi, Vec<u8>)>,
+    ) -> Self {
+        let num_of_optimizations =
+            if config.optimizer { Some(config.optimizer_runs) } else { None };
```
Comment (gakonst) [inline cli/src/cmd/forge/script/mod.rs:450]: Should we not be getting this from the cache file?

(not for this PR anyway)

File: cli/src/cmd/forge/script/sequence.rs (hunk 11 above)
Comment (gakonst) [reply cli/src/cmd/forge/script/sequence.rs:135]: Looks good - but let's also pull this in a helper function on `Libraries`

**Review by gakonst — APPROVED** [review]
beautiful

**Comment by gakonst** [comment]
I kinda don't want to bundle the ABI to the output file, but can see the argument for it, especially if people want a 1-click instantiation of a contract(s).

**Comment by mds1** [comment]
Sorry missed this one before it got merged, overall love the format. Some comments:

1. We might want to rename the `type` key to `kind` or something similar—`type` is a keyword in many languages so it's a great key name
2. The function signatures for CALLs look like this: `"function": "inc():(uint256)"`. What is that format, specifically why is there a colon separating the input and output args? I'd expect to see `"function": "inc()(uint256)"`

**Comment by gakonst** [comment]
Agree with all feedback points!

**Comment by ShravanSunder** [comment]
is this now part of the currently nightly?

**Comment by gakonst** [comment]
Should be - yah

**Pull request merged**